Userland/Libraries/LibGUI/TextDocument.cpp

```diff
@@ -383,6 +407,96 @@ DeprecatedString TextDocument::text_in_range(TextRange const& a_range) const
+size_t TextDocument::get_code_points_before_cursor(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    return get_prev_grapheme_cluster_boundary(cursor);
+}
+
+// This function will return the position of the next grapheme cluster break, relative to the cursor, for "correct looking" parsing of unicode based on grapheme cluster boundary algorithm.
+size_t TextDocument::get_next_grapheme_cluster_boundary(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    if (cursor.column() == line.length())
+        return 0;
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(Utf32View(line.code_points(), line.length()));
+
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    if (boundaries_cache.size() == 1)
+        return cursor.column() - boundaries_cache.at(0);
+
+    auto previous_boundary_guess = boundaries_cache.at(0);
+    auto boundary_guess = boundaries_cache.at(1);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the left, update our guess
+        if (current_boundary_guess < boundary_guess) {
+            previous_boundary_guess = boundary_guess;
+            boundary_guess = current_boundary_guess;
+        }
+    }
+
+    // Assume that the cursor cannot sit at a point that is not a boundary, so return previous guess.
+    return previous_boundary_guess;
+}
+
+// This function will return the position of the previous grapheme cluster break, relative to the cursor, for "correct looking" parsing of unicode based on grapheme cluster boundary algorithm.
+size_t TextDocument::get_prev_grapheme_cluster_boundary(TextPosition const& cursor) const
+{
+    if (!cursor.is_valid())
+        return 0;
+
+    auto& line = this->line(cursor.line());
+    auto line_span = Span<u32 const>(line.code_points(), line.length());
+
+    if (line_span.is_empty())
+        return 0;
+
+    if (cursor.column() < 2)
+        return cursor.column();
+
+    auto boundaries_cache = Unicode::find_grapheme_segmentation_boundaries(Utf32View(line.code_points(), line.length()));
+    if (boundaries_cache.is_empty())
+        return cursor.column();
+
+    if (boundaries_cache.size() == 1)
+        return cursor.column() - boundaries_cache.at(0);
+
+    auto boundary_guess = boundaries_cache.at(0);
+
+    for (auto current_boundary_guess : boundaries_cache) {
+        // If the current boundary is further to the right, update our guess.
+        if ((current_boundary_guess - cursor.column()) > (boundary_guess - cursor.column()))
+            boundary_guess = current_boundary_guess;
+    }
+
+    // Return the distance in code points from the cursor.
+    return cursor.column() - boundary_guess;
```

Comment:
I would expect this function to just return the `boundary_guess` - i.e. it should find the boundary, not the distance between the boundary and the cursor. Let the caller (`get_code_points_before_cursor`) handle that.